node/src/components/fetcher/trie_fetcher.rs

```diff
@@ -0,0 +1,342 @@
+                        );
+                        return partial_chunks
+                            .responder
+                            .respond(Err(TrieFetcherError::Absent(digest, index)))
+                            .ignore();
+                    }
+                };
+                let next_id = TrieOrChunkedDataId(missing_index, digest);
+                self.try_download_chunk(effect_builder, next_id, peer, partial_chunks)
+            }
+            None => match partial_chunks.assemble_chunks(count) {
+                Ok(trie) => partial_chunks
+                    .responder
+                    .respond(into_response(Box::new(trie), partial_chunks.sender))
+                    .ignore(),
+                Err(error) => {
+                    error!(%digest, %error,
+                        "error while assembling a complete trie",
+                    );
+                    partial_chunks.responder.respond(Err(error.into())).ignore()
+                }
+            },
+        }
+    }
+
+    fn try_download_chunk<REv>(
+        &mut self,
+        effect_builder: EffectBuilder<REv>,
+        id: TrieOrChunkedDataId,
+        peer: I,
+        partial_chunks: PartialChunks<I>,
+    ) -> Effects<Event<I>>
+    where
+        REv: ReactorEventT<TrieOrChunkedData> + From<FetcherRequest<I, TrieOrChunkedData>>,
+    {
+        let TrieOrChunkedDataId(_, hash) = id;
+        let _ = self.partial_chunks.insert(hash, partial_chunks);
+        effect_builder
+            .fetch(id, peer)
+            .event(move |fetch_result| Event::TrieOrChunkFetched { id, fetch_result })
+    }
+}
+
+impl<I, REv> Component<REv> for TrieFetcher<I>
+where
+    REv: ReactorEventT<TrieOrChunkedData> + From<FetcherRequest<I, TrieOrChunkedData>>,
+    I: Debug + Clone + Send + Eq + 'static,
+{
+    type Event = Event<I>;
+    type ConstructionError = prometheus::Error;
+
+    fn handle_event(
+        &mut self,
+        effect_builder: EffectBuilder<REv>,
+        _rng: &mut NodeRng,
+        event: Self::Event,
+    ) -> Effects<Self::Event> {
+        debug!(?event, "handling event");
+        match event {
+            Event::Request(TrieFetcherRequest {
+                hash,
+                responder,
+                peers,
+            }) => {
+                let trie_id = TrieOrChunkedDataId(0, hash);
+                let peer = match peers.last() {
+                    Some(peer) => peer.clone(),
+                    None => {
+                        error!(%hash, "tried to fetch trie with no peers available");
+                        return Effects::new();
+                    }
+                };
+                let partial_chunks = PartialChunks {
+                    responder,
+                    peers,
+                    chunks: Default::default(),
+                    sender: None,
+                };
+                self.try_download_chunk(effect_builder, trie_id, peer, partial_chunks)
```

Comment:
The problem is that I chose to react to the lack of peers slightly differently depending on the stage of the process, so it would be a bit hard to do that.